File: scalaz/src/main/scala/env.scala
import types._

import scala.collection.mutable

object env {

  class Env private(outer: Option[Env], binds: Iterator[MalSymbol], exprs: Iterator[MalType]) {
    private val data: mutable.Map[MalSymbol, MalType] = mutable.Map.empty
    for (bind <- binds) {
      bind match {
        case MalSymbol.sp.Variadic =>
          val symbol = binds.next()
          val expr = MalList(exprs.toList)
          data(symbol) = expr
        case _ =>
          data(bind) = exprs.next()
      }
    }

    def apply(symbol: MalSymbol): Option[MalType] = for (env <- find(symbol)) yield env.data(symbol)

    def update(symbol: MalSymbol, value: MalType): Unit = data(symbol) = value

    def find(symbol: MalSymbol): Option[Env] = data.get(symbol) match {
      case Some(_) => Some(this)
      case None => outer.flatMap(_.find(symbol))
    }

    def inner(binds: Seq[MalSymbol] = Nil, exprs: Seq[MalType] = Nil): Env =
      new Env(Some(this), binds.iterator, exprs.iterator)
  }

  object Env {
    def apply(binds: Seq[MalSymbol] = Nil, exprs: Seq[MalType] = Nil): Env =
      new Env(None, binds.iterator, exprs.iterator)
    implicit def toSymbol(symbol: Symbol): MalSymbol = MalSymbol(symbol)
  }

}

File: jvzgo/src/env/env.go
package env

import (
	"fmt"
	. "types"
)

type Env struct {
	outer EnvType
	data  map[string]MalType
}

func NewEnv() EnvType {
	return &Env{data: make(map[string]MalType)}
}

func (env *Env) New(binds, exprs []MalType) (EnvType, error) {
	inner := Env{outer: env, data: make(map[string]MalType)}
	for i := 0; i < len(binds); i++ {
		sym, err := GetSymbol(binds[i])
		if err != nil {
			return nil, err
		}
		if sym.Value == "&" {
			sym, err := GetSymbol(binds[i+1])
			if err != nil {
				return nil, err
			}
			inner.Set(sym.Value, NewList(exprs[i:]))
			break
		}
		inner.Set(sym.Value, exprs[i])
	}
	return &inner, nil
}

func (env *Env) Set(key string, val MalType) {
	if env.data == nil {
		env.data = make(map[string]MalType)
	}
	env.data[key] = val
}

func (env *Env) Find(key string) EnvType {
	if _, ok := env.data[key]; ok {
		return env
	} else if env.outer != nil {
		return env.outer.Find(key)
	} else {
		return nil
	}
}

func (env *Env) Get(key string) (MalType, error) {
	e := env.Find(key)
	if e == nil {
		return nil, fmt.Errorf("unknown key: %v", key)
	}
	return e.(*Env).data[key], nil
}
